File: 1306-90-1765-map-of-highest-peak/1306-90-1765-map-of-highest-peak.java

class Solution {
    public int[][] highestPeak(int[][] isWater) {
        int n = isWater.length, m = isWater[0].length;
        boolean[][] visited = new boolean[n][m];
        Queue<int[]> q = new LinkedList<>();
        
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (isWater[i][j] == 1) {
                    isWater[i][j] = 0;
                    q.add(new int[]{i, j});
                    visited[i][j] = true;
                }
            }
        }
        
        int[] directions = {-1, 0, 1, 0, -1};
        
        while (!q.isEmpty()) {
            int[] cell = q.poll();
            int x = cell[0], y = cell[1];
            for (int i = 0; i < 4; i++) {
                int dx = x + directions[i];
                int dy = y + directions[i + 1];
                if (dx >= 0 && dy >= 0 && dx < n && dy < m && !visited[dx][dy]) {
                    visited[dx][dy] = true;
                    isWater[dx][dy] = isWater[x][y] + 1;
                    q.add(new int[]{dx, dy});
                }
            }
        }
        
        return isWater;
    }
}

File: 1306-90-1765-map-of-highest-peak/1306-90-1765-map-of-highest-peak.py
from collections import deque

class Solution:
    def highestPeak(self, isWater: list[list[int]]) -> list[list[int]]:
        n, m = len(isWater), len(isWater[0])
        visited = [[False] * m for _ in range(n)]
        q = deque()
        
        for i in range(n):
            for j in range(m):
                if isWater[i][j] == 1:
                    isWater[i][j] = 0
                    q.append((i, j))
                    visited[i][j] = True
        
        directions = [-1, 0, 1, 0, -1]
        
        while q:
            x, y = q.popleft()
            for i in range(4):
                dx, dy = x + directions[i], y + directions[i + 1]
                if 0 <= dx < n and 0 <= dy < m and not visited[dx][dy]:
                    visited[dx][dy] = True
                    isWater[dx][dy] = isWater[x][y] + 1
                    q.append((dx, dy))
        
        return isWater
